### PR TITLE
[storage] add proptest for jellyfish internal node

### PR DESCRIPTION
**Pull request by msmouse**
## Motivation

To cover all (valid node) cases.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

proptest added.

### CODE REVIEW

File: storage/jellyfish_merkle/src/node_type/mod.rs [hunk 1]
```diff
@@ -268,6 +269,23 @@ impl From<LeafNode> for Node {
...
+                "InternalNode with only one child which is a leaf is illegal",
+                |children| {
+                    !(children.len() == 1 || children.values().next().expect("Must exist.").is_leaf)
+                },
+            )
+            .prop_map(InternalNode::new)
+            .boxed()
+    }
+
+    type Strategy = BoxedStrategy<Self>;
```
Comment (wqfish) [inline storage/jellyfish_merkle/src/node_type/mod.rs]: nit: move this up?

```
type Parameters = ();
type Strategy = BoxedStrategy<Self>;

fn arbitrary_with(...)
```

File: storage/jellyfish_merkle/src/node_type/mod.rs [hunk 2]
```diff
@@ -268,6 +269,23 @@ impl From<LeafNode> for Node {
     }
 }
 
+impl Arbitrary for InternalNode {
+    type Parameters = ();
+    fn arbitrary_with(_args: ()) -> Self::Strategy {
+        hash_map(any::<Nibble>(), any::<Child>(), 1..=16)
+            .prop_filter(
+                "InternalNode with only one child which is a leaf is illegal",
```
Comment (wqfish) [inline storage/jellyfish_merkle/src/node_type/mod.rs]:
```suggestion
                "InternalNode with only one leaf child is illegal",
```

File: storage/jellyfish_merkle/src/node_type/mod.rs [hunk 3]
```diff
@@ -268,6 +269,23 @@ impl From<LeafNode> for Node {
...
+                |children| {
+                    !(children.len() == 1 || children.values().next().expect("Must exist.").is_leaf)
```
Comment (wqfish) [inline storage/jellyfish_merkle/src/node_type/mod.rs]: Is this `||` or `&&`?

File: storage/jellyfish_merkle/src/node_type/node_type_test.rs [hunk 4]
```diff
@@ -669,3 +669,162 @@ fn test_internal_hash_and_proof() {
...
+        assert!(n < self.first_child_index + self.num_children);
+
+        n < self.first_child_index + self.num_children / 2
+    }
+}
+
+#[derive(Clone, Copy)]
+struct BinaryTreeChildNode {
+    version: Version,
+    index: u8,
```
Comment (wqfish) [inline storage/jellyfish_merkle/src/node_type/node_type_test.rs]: Would be good to add some comments for these fields. It takes some time to figure out what `first_child_index`, `num_children`, `index` are.

File: storage/jellyfish_merkle/src/node_type/node_type_test.rs [hunk 5]
```diff
@@ -669,3 +669,162 @@ fn test_internal_hash_and_proof() {
...
+    hash: HashValue,
+    is_leaf: bool,
+}
+
+struct NaiveInternalNode {
+    root: Rc<BinaryTreeNode>,
+}
+
+impl NaiveInternalNode {
+    fn from_clever_node(node: &InternalNode) -> Self {
```
Comment (lightmark) [inline storage/jellyfish_merkle/src/node_type/node_type_test.rs]: clever?

File: storage/jellyfish_merkle/src/node_type/node_type_test.rs (hunk 4 above)
Comment (lightmark) [reply storage/jellyfish_merkle/src/node_type/node_type_test.rs]: confused about `index` too

File: storage/jellyfish_merkle/src/node_type/node_type_test.rs [hunk 6]
```diff
@@ -669,3 +669,162 @@ fn test_internal_hash_and_proof() {
...
+        Self {
+            root: Rc::new(Self::node_for_subtree(0, 16, &node.children)),
+        }
+    }
+
+    fn node_for_subtree(
+        first_child_index: u8,
+        num_children: u8,
```
Comment (lightmark) [inline storage/jellyfish_merkle/src/node_type/node_type_test.rs]: This name is confusing.

Comment (msmouse) [reply storage/jellyfish_merkle/src/node_type/node_type_test.rs]: see latest commit :P

File: storage/jellyfish_merkle/src/node_type/node_type_test.rs (hunk 5 above)
Comment (msmouse) [reply storage/jellyfish_merkle/src/node_type/node_type_test.rs]: because this is `naive`..

File: storage/jellyfish_merkle/src/node_type/mod.rs [hunk 7]
```diff
@@ -271,26 +271,26 @@ impl From<LeafNode> for Node {
 
 impl Arbitrary for InternalNode {
     type Parameters = ();
+    type Strategy = BoxedStrategy<Self>;
+
     fn arbitrary_with(_args: ()) -> Self::Strategy {
         hash_map(any::<Nibble>(), any::<Child>(), 1..=16)
             .prop_filter(
-                "InternalNode with only one child which is a leaf is illegal",
+                "InternalNode with only one leaf child is illegal",
```
Comment (msmouse) [inline storage/jellyfish_merkle/src/node_type/mod.rs]: o this sounds not right.. >_< will rephrase again.

Comment (zamsden) [reply storage/jellyfish_merkle/src/node_type/mod.rs]: "child leaf" is better

Comment (msmouse) [reply storage/jellyfish_merkle/src/node_type/mod.rs]: "InternalNode constructor panics when its only child is a leaf."

how about this

**Comment by wqfish** [comment]
The commits weren't squashed. Maybe you should use `git commit --fixup` or manual squash?

**Pull request closed without merging**